external/java/src/main/java/ch/epfl/dedis/lib/omniledger/contracts/DarcInstance.java

```diff
@@ -0,0 +1,159 @@
+import ch.epfl.dedis.lib.omniledger.*;
+import ch.epfl.dedis.lib.omniledger.darc.*;
+import com.google.protobuf.InvalidProtocolBufferException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.xml.bind.DatatypeConverter;
+import java.util.Arrays;
+
+/**
+ * DarcInstance represents an instance of a darc on Omniledger. It is self-
+ * sufficient, meaning it has a link to the omniledger instance it runs on.
+ * If you evolve the DarcInstance, it will update its internal darc.
+ */
+public class DarcInstance {
+    private Instance instance;
+    private Darc darc;
+    private OmniledgerRPC ol;
+
+    private final static Logger logger = LoggerFactory.getLogger(DarcInstance.class);
+
+    /**
+     * Instantiates a new DarcInstance given a working omniledger instance and
+     * an instanceId. This instantiator will contact omniledger and try to get
+     * the current darcInstance. If the instance is not found, or is not of
+     * contractId "Darc", an exception will be thrown.
+     *
+     * @param ol is a link to an omniledger instance that is running
+     * @param id of the darc-instance to connect to
+     * @throws CothorityException
+     */
+    public DarcInstance(OmniledgerRPC ol, InstanceId id) throws CothorityException {
+        this.ol = ol;
+        Proof p = ol.getProof(id);
+        instance = new Instance(p);
+        if (!instance.getContractId().equals("darc")) {
+            logger.error("wrong instance: {}", instance.getContractId());
+            throw new CothorityNotFoundException("this is not a darc instance");
+        }
+        try {
+            darc = new Darc(instance.getData());
+        } catch (InvalidProtocolBufferException e) {
+            throw new CothorityCryptoException(e.getMessage());
+        }
+    }
+
+    public DarcInstance(OmniledgerRPC ol, Darc d) throws CothorityException {
+        this(ol, new InstanceId(d.getBaseId(), SubId.zero()));
+    }
+
+    public void update() throws CothorityException {
+        instance = new Instance(ol.getProof(instance.getId()));
+        try {
+            darc = new Darc(instance.getData());
+        } catch (InvalidProtocolBufferException e) {
+            throw new CothorityCryptoException(e.getMessage());
+        }
+    }
+
+    /**
+     * Creates an instruction to evolve the darc in omniledger. The signer must have its identity in the current
+     * darc as "Invoke_Evolve" rule.
+     *
+     * @param newDarc the darc to replace the old darc.
+     * @param owner   must have its identity in the "Invoke_Evolve" rule
+     * @param pos     position of the instruction in the ClientTransaction
+     * @param len     total number of instructions in the ClientTransaction
+     * @return Instruction to be sent to omniledger
+     * @throws CothorityCryptoException
+     */
+    public Instruction evolveDarcInstruction(Darc newDarc, Signer owner, int pos, int len) throws CothorityCryptoException {
```

Comment:
s/owner/signer/ I think. Because any authorized signer can ask for the evolution, not just the owner. (The default is owner only, but the owner could give others evolution rights?)

Reply:
My definition of `owner` is any set of identities that can resolve the expression in the `invoke:evolve` rule. I added a TODO but kept `owner`.